pkg/utils/metadata.go

```diff
@@ -17,31 +17,51 @@ limitations under the License.
 package utils
 
 import (
+	"context"
+	"os"
+
 	metadataClient "github.com/google/knative-gcp/pkg/gclient/metadata"
 )
 
 const (
 	clusterNameAttr = "cluster-name"
+	// ProjectIDEnvKey is the name of environmental variable for project ID
 	ProjectIDEnvKey = "PROJECT_ID"
 )
 
+// defaultMetadataClientCreator is a create function to get a default metadata client. This can be
+// swapped during testing.
+var defaultMetadataClientCreator func() metadataClient.Client = metadataClient.NewDefaultMetadataClient
+
+// projectIDFromEnv loads project ID from env once when startup.
+var projectIDFromEnv string
+
+func init() {
+	projectIDFromEnv = os.Getenv(ProjectIDEnvKey)
+}
+
 // ProjectIDEnvConfig is a struct to parse project ID from env var
 type ProjectIDEnvConfig struct {
 	ProjectID string `envconfig:"PROJECT_ID"`
 }
 
-// ProjectID returns the project ID for a particular resource.
-func ProjectID(project string, client metadataClient.Client) (string, error) {
-	// If project is set, then return that one.
-	if project != "" {
-		return project, nil
+// ProjectIDOrDefault returns the project ID by performing the following order:
+// 1) if the input project ID is valid, simply use it.
+// 2) if there is a PROJECT_ID environmental variable, use it.
+// 3) use metadataClient to resolve project id.
+func ProjectIDOrDefault(ctx context.Context, projectID string) (string, error) {
+	if projectID != "" {
+		return projectID, nil
+	}
+	if projectIDFromEnv != "" {
+		return projectIDFromEnv, nil
```

Comment:
It seems we didn't have unit test for function `init()`(maybe I missed that). Did you manually test, if env `PROJECT_ID` has value (assume it has a different value than what would get from  defaultMetadataClientCreator()), the function` init()` will parse that in?

Reply:
Not sure if I understand the question here. We didn't test `init()` in unit test, though I actually test it manually to confirm `os.Getenv` works. The `init` is called only when the package is loaded, so after that, we will simply use the `projectIDFromEnv` directly and don't read the env again.

Reply:
It looks like `spec.project` is translated into the `PROJECT_ID` env var for the receive adapter. So if there's an existing E2E test that verifies `spec.project` gets down to the receive adapter, that would also verify the operation of the `init()` function that reads `PROJECT_ID`.

Reply:
@zhongduo The question is for verifying the operation of the `init()` function. 
@grantr Thank you for reminding this!
In most of the case, the env `project_id` maybe the same whether we get it from os.Getenv, or from the client. Just want to make sure we manually test that: when we set `project_id` env, it returns the value from `init()`, not from the client.

Reply:
I see. Yes, i manually tested it by commenting out the line setting `projectIDFromEnv` in unit test, and the test now gets my `PROJECT_ID` env var.

Reply:
@capri-xiyue do we have an E2E test that verifies the `spec.project` field in sources works correctly?

Reply:
@grantr No, we don't. In addition, the idea of have `project` in spec is to support cross-project stuff as Nacho mentioned in this comment https://github.com/google/knative-gcp/pull/1761#discussion_r496937485. It may be difficult to do e2e test for cross-project stuff in prow. If we just verify `spec.project` when the `spec.project` is the same as the default project of gke server, the e2e test seems useless.

Reply:
Most of the unit tests in sources rely no project field in the spec, so I think that feature is well tested. And the logic currently is: if project is not set, we use the default project. That's why you see there are lots of `ProjectIDOrDefault` calls in sources.

Reply:
Ok, sounds like a manual test of the init behavior is fine for now.